src/Microsoft.Extensions.Logging.AzureAppServices/Internal/BatchingLoggerProvider.cs

```diff
@@ -0,0 +1,164 @@
+        private readonly int? _batchSize;
+        private readonly IDisposable _optionsChangeToken;
+
+        private BlockingCollection<LogMessage> _messageQueue;
+        private Task _outputTask;
+        private CancellationTokenSource _cancellationTokenSource;
+
+        protected BatchingLoggerProvider(IOptionsMonitor<BatchingLoggerOptions> options)
+        {
+            // NOTE: Only IsEnabled is monitored
+
+            var loggerOptions = options.CurrentValue;
+            if (loggerOptions.BatchSize < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(loggerOptions.BatchSize), $"{nameof(loggerOptions.BatchSize)} must be a positive number.");
+            }
+            if (loggerOptions.FlushPeriod <= TimeSpan.Zero)
+            {
+                throw new ArgumentOutOfRangeException(nameof(loggerOptions.FlushPeriod), $"{nameof(loggerOptions.FlushPeriod)} must be longer than zero.");
+            }
+
+            _interval = loggerOptions.FlushPeriod;
+            _batchSize = loggerOptions.BatchSize;
+            _queueSize = loggerOptions.BackgroundQueueSize;
+
+            _optionsChangeToken = options.OnChange(UpdateOptions);
+            UpdateOptions(options.CurrentValue);
+        }
+
+        public bool IsEnabled { get; private set; }
+
+        private void UpdateOptions(BatchingLoggerOptions options)
+        {
+            var oldIsEnabled = IsEnabled;
+            IsEnabled = options.IsEnabled;
+            if (oldIsEnabled != IsEnabled)
+            {
+                if (IsEnabled)
+                {
+                    Start();
+                }
+                else
+                {
+                    Stop();
+                }
+            }
+
+        }
+
+        protected abstract Task WriteMessagesAsync(IEnumerable<LogMessage> messages, CancellationToken token);
+
+        private async Task ProcessLogQueue(object state)
+        {
+            while (!_cancellationTokenSource.IsCancellationRequested)
+            {
+                var limit = _batchSize ?? int.MaxValue;
+
+                while (limit > 0 && _messageQueue.TryTake(out var message))
+                {
+                    _currentBatch.Add(message);
+                    limit--;
+                }
+
+                if (_currentBatch.Count > 0)
+                {
+                    try
+                    {
+                        await WriteMessagesAsync(_currentBatch, _cancellationTokenSource.Token);
+                    }
+                    catch
+                    {
+                        // ignored
+                    }
+
+                    _currentBatch.Clear();
+                }
+
+                await IntervalAsync(_interval, _cancellationTokenSource.Token);
+            }
+        }
+
+        protected virtual Task IntervalAsync(TimeSpan interval, CancellationToken cancellationToken)
+        {
+            return Task.Delay(interval, cancellationToken);
```

Comment:
Nit/personal preference: expression body syntax can be used here to reduce this to one line. We aren't consistent on this, but IMO we should use it where possible.

This is true of several dozen places through out this PR, so I'll only mention it here.